File: backend/app/schemas/software.py
"""Pydantic schemas for Software API request/response validation."""

from datetime import date, datetime
from decimal import Decimal
from typing import Optional, List
from pydantic import BaseModel, Field, ConfigDict


class SoftwareBase(BaseModel):
    """Base schema with common software fields."""
    category: Optional[str] = Field(None, max_length=100)
    name: str = Field(..., max_length=200)
    version: Optional[str] = Field(None, max_length=50)
    key: Optional[str] = Field(None, max_length=500)
    type: Optional[str] = Field(None, max_length=100)
    purchase_date: Optional[date] = None
    purchaser: Optional[str] = Field(None, max_length=200)
    vendor: Optional[str] = Field(None, max_length=200)
    cost: Optional[Decimal] = Field(None, ge=0)
    deployment: Optional[str] = Field(None, max_length=200)
    install_location: Optional[str] = Field(None, max_length=500)
    status: Optional[str] = Field(None, max_length=50)
    comments: Optional[str] = None


class SoftwareCreate(SoftwareBase):
    """Schema for creating new software."""
    pass


class SoftwareUpdate(BaseModel):
    """Schema for updating software (all fields optional)."""
    category: Optional[str] = Field(None, max_length=100)
    name: Optional[str] = Field(None, max_length=200)
    version: Optional[str] = Field(None, max_length=50)
    key: Optional[str] = Field(None, max_length=500)
    type: Optional[str] = Field(None, max_length=100)
    purchase_date: Optional[date] = None
    purchaser: Optional[str] = Field(None, max_length=200)
    vendor: Optional[str] = Field(None, max_length=200)
    cost: Optional[Decimal] = Field(None, ge=0)
    deployment: Optional[str] = Field(None, max_length=200)
    install_location: Optional[str] = Field(None, max_length=500)
    status: Optional[str] = Field(None, max_length=50)
    comments: Optional[str] = None


class SoftwareListItem(BaseModel):
    """Summary view of software for list display."""
    id: int
    software_id: str
    category: Optional[str] = None
    name: str
    version: Optional[str] = None
    key: Optional[str] = None
    type: Optional[str] = None
    purchase_date: Optional[date] = None
    purchaser: Optional[str] = None
    vendor: Optional[str] = None
    cost: Optional[Decimal] = None
    deployment: Optional[str] = None
    install_location: Optional[str] = None
    status: Optional[str] = None
    comments: Optional[str] = None
    is_deleted: bool

    model_config = ConfigDict(from_attributes=True)


class SoftwareResponse(SoftwareBase):
    """Full software record response."""
    id: int
    software_id: str
    created_at: datetime
    updated_at: datetime
    is_deleted: bool
    deleted_at: Optional[datetime] = None

    model_config = ConfigDict(from_attributes=True)


class SoftwareImportError(BaseModel):
    """Details of a failed import row."""
    row: int
    field: Optional[str] = None
    message: str


class SoftwareImportResult(BaseModel):
    """Result of CSV import operation."""
    success_count: int
    error_count: int
    errors: List[SoftwareImportError]

File: frontend/src/types/software.ts
// Software types matching OpenAPI schemas

// Predefined values for dropdowns
export const SOFTWARE_CATEGORIES = ['Office', 'Security', 'Development', 'Business', 'Utilities'] as const;
export const SOFTWARE_TYPES = ['Subscription', 'Perpetual', 'Volume', 'OEM', 'Freeware', 'Open Source'] as const;
export const SOFTWARE_STATUSES = ['Active', 'Inactive', 'Expired', 'Retired'] as const;

// List item for software table
export interface SoftwareListItem {
  id: number;
  software_id: string;
  category: string | null;
  name: string;
  version: string | null;
  key: string | null;
  type: string | null;
  purchase_date: string | null;
  purchaser: string | null;
  vendor: string | null;
  cost: number | null;
  deployment: string | null;
  install_location: string | null;
  status: string | null;
  comments: string | null;
  is_deleted: boolean;
}

// Full software record
export interface Software {
  id: number;
  software_id: string;
  category: string | null;
  name: string;
  version: string | null;
  key: string | null;
  type: string | null;
  purchase_date: string | null;
  purchaser: string | null;
  vendor: string | null;
  cost: number | null;
  deployment: string | null;
  install_location: string | null;
  status: string | null;
  comments: string | null;
  created_at: string;
  updated_at: string;
  is_deleted: boolean;
  deleted_at: string | null;
}

// Request body for creating software
export interface SoftwareCreate {
  name: string;
  category?: string;
  version?: string;
  key?: string;
  type?: string;
  purchase_date?: string;
  purchaser?: string;
  vendor?: string;
  cost?: number;
  deployment?: string;
  install_location?: string;
  status?: string;
  comments?: string;
}

// Request body for updating software
export interface SoftwareUpdate {
  name?: string;
  category?: string;
  version?: string;
  key?: string;
  type?: string;
  purchase_date?: string;
  purchaser?: string;
  vendor?: string;
  cost?: number;
  deployment?: string;
  install_location?: string;
  status?: string;
  comments?: string;
}

// Import result
export interface SoftwareImportResult {
  success_count: number;
  error_count: number;
  errors: SoftwareImportError[];
}

export interface SoftwareImportError {
  row: number;
  field: string | null;
  message: string;
}

// Filter parameters
export interface SoftwareFilters {
  search?: string;
  category?: string;
  status?: string;
  type?: string;
  vendor?: string;
  purchaser?: string;
  deployment?: string;
  sort_by?: 'software_id' | 'category' | 'name' | 'version' | 'type' | 'vendor' | 'status' | 'purchase_date' | 'created_at';
  sort_order?: 'asc' | 'desc';
  include_deleted?: boolean;
}
